### PR TITLE
Require export before delete in VxScan

### PR DESCRIPTION
**Pull request by taravancil**
## Overview
#1384 

This does several things:

* Adds a table `backups` that tracks timestamps of given backup keys (currently `Cvrs` and `Batches`)
* Adds `store.getCanUnconfigure()`, which determines whether the machine can be unconfigured
* Updates `DELETE /config/election` and `POST /scan/zero` to require `store.getCanUnconfigure()` to return true.
* Updates the VxScan admin UI to disable the "Delete Election Data..." button until CVRs and batches have been backed up. The button is disabled again when an export is invalidated by either adding or deleting a batch or adding new sheets. The button is always enabled in test mode.
  * _Note that a machine with no batches will allow deleting election data. This seems to me like the right thing to do, but I am open to other opinions_

The work to implement this in VxCentralScan will happen in a separate ticket.

## Demo Video or Screenshot
Uploading Screen Recording 2022-04-14 at 5.53.33 AM.mov…

## Testing Plan
I have added several tests and manually QAed this flow.

## Checklist
- [x] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.
- [x] I have added JSDoc comments to any newly introduced exports

### CODE REVIEW

File: libs/utils/src/date.ts
Comment (eventualbuddha) [inline libs/utils/src/date.ts:81]: `eslint` says:
> Missing return type on function. `eslint@typescript-eslint/explicit-module-boundary-types`

File: libs/utils/src/date.ts
Comment (eventualbuddha) [inline libs/utils/src/date.ts:83]: Maybe just use `Intl.DateTimeFormat` for this?

File: frontends/bsd/src/app_root.tsx
Comment (eventualbuddha) [inline frontends/bsd/src/app_root.tsx:646]: Could be abbreviated:
```suggestion
              exportedAt={status.batches[0]?.exportedAt}
```
No need for the `''` as `undefined` should suffice.

File: services/scan/src/store.ts
Comment (eventualbuddha) [inline services/scan/src/store.ts:645]: Does this really invalidate anything? Any previous backups would still have all the other batches (plus this one). I'm unsure how this ought to work cc @benadida.

File: frontends/bsd/src/screens/admin_actions_screen.tsx
Comment (eventualbuddha) [inline frontends/bsd/src/screens/admin_actions_screen.tsx:27]:
```suggestion
  exportedAt?: Iso8601Timestamp
```

File: services/scan/src/server.ts
Comment (eventualbuddha) [inline services/scan/src/server.ts:166]: This construct is used a few times in this PR. Consider making something equivalent to this a method on `Store`.

**Review by eventualbuddha — COMMENTED** [review]
A good start! Do we also want to prevent unconfigure until CVRs have been exported? cc @benadida

**Comment by benadida** [comment]
> A good start! Do we also want to prevent unconfigure until CVRs have been exported? cc @benadida

Yes, that's a good call. A few enhancements here that should have been part of the ticket from the start, my bad for not specifying them:

- we should require CVR export.
- we should block "Delete Ballot Data" in the same way, and we can put that button at the bottom so it's grouped together with the big red button. In fact I'd say both those buttons should be red.
- we should NOT block if we're in test mode – it's just test ballots and the warning is fine.

@taravancil the UX / text is good by me. And sorry for the added scope here, I should have scoped this ticket better.

File: services/scan/src/store.ts
Comment (taravancil) [reply services/scan/src/store.ts:449]: That is how it worked before d2fe24b0223dc1277930a73dc0a919dbffaf06bb, but as I was working on invalidating the export when _deleting a batch_, I realized the approach of marking individual batches with `exported_at=null` would not work in that case, because each remaining batch would still have the `exported_at` column set.

As an alternative approach, we could use a single flag `exported_at` instead of `exported_at` on each batch. That is effectively how `batches.exported_at` works in this current state, anyway.

I guess before going any further, I should also confirm that deleting a batch should invalidate the export. What do you think @eventualbuddha? If not, I can revert d2fe24b0223dc1277930a73dc0a919dbffaf06bb and go back to individually invalidating batches.

File: services/scan/src/store.ts
Comment (taravancil) [reply services/scan/src/store.ts:645]: Yeah, I was unsure about whether deleting a batch should invalidate an export. As mentioned [here](https://github.com/votingworks/vxsuite/pull/1580#discussion_r827341485), if this should not invalidate the backup, we can revert d2fe24b0223dc1277930a73dc0a919dbffaf06bb.

**Comment by taravancil** [comment]
> we should require CVR export.

@benadida checking my understanding, the resulting .zip when I click "Export Backup..." includes a `cvrs.jsonl` file. ([source](https://github.com/votingworks/vxsuite/blob/7d6b73ba5eb117cd28c1ac817b8d89ec765938e7/services/scan/src/backup.ts#L73,L75)). Is that sufficient or is there more to do here?

**Comment by benadida** [comment]
> > we should require CVR export.
> 
> @benadida checking my understanding, the resulting .zip when I click "Export Backup..." includes a `cvrs.jsonl` file. ([source](https://github.com/votingworks/vxsuite/blob/7d6b73ba5eb117cd28c1ac817b8d89ec765938e7/services/scan/src/backup.ts#L73,L75)). Is that sufficient or is there more to do here?

Not sufficient, cause the election admin will use those two files differently.

File: frontends/bsd/src/screens/admin_actions_screen.tsx
Comment (beausmith) [inline frontends/bsd/src/screens/admin_actions_screen.tsx:171]: The text above the button is equally close to the button above it too. We should more closely associate the text to the button it's related to.

I haven't run the code below to ensure it works, but it should:
- Place the text to the right of the button.
- Update copy to be more specific. (Though now realizing that "Export Backup" should probably be changed to something more descriptive which doesn't contain two closely related verbs/tems. 🙃 )
- Use warning style text.
- Use a warning icon.

I've removed the italic style as I'd like to save that for adding emphasis to content next to non-italicized content.

```suggestion
                <Button
                  danger
                  disabled={!hasExportedBackupForAllBatches}
                  onPress={toggleIsConfirmingUnconfigure}
                >
                  Delete Election Data from VxCentralScan…
                </Button>{' '}
                {!hasExportedBackupForAllBatches && (
                  <Text as="span" warning warningIcon>
                    You must "Export Backup" before you may delete election data.
                  </Text>
                )}
                {hasExportedBackupForAllBatches && exportedAt && (
                  <Text as="span" muted>
                    Election data was last backed up at {shortDateTime(exportedAt)}
                  </Text>
                )}
```

File: frontends/bsd/src/screens/admin_actions_screen.tsx
Comment (beausmith) [inline frontends/bsd/src/screens/admin_actions_screen.tsx:163]: Related to other comment.

```suggestion
```

**Review by beausmith — APPROVED** [review]
This looks good. I've got a suggestion for the visual display. LMK if you feel otherwise.

File: services/scan/src/server.ts
Comment (taravancil) [reply services/scan/src/server.ts]: I've added `Store.getCanUnconfigure()`, and then also added `canUnconfigure` to `GetScanStatusResponse`.

**Comment by taravancil** [comment]
@benadida called out that manually invalidating the backup when we add/delete batches/sheets is brittle, because it requires us to remember to invalidate backups if/when we add new features. He made a good suggestion that we could compare export timestamps to the max of `batches.ended_at`, `sheets.created_at`, `sheets.deleted_at` to decide if a backup is up-to-date. I like this idea, but it still requires manually invalidating the export when we delete a batch.

We could get around this by setting `deleted_at` on batches instead of actually deleting batches (like we do for sheets), but before I move forward with that implementation, I wanted to pause and see if 1. we think it's worth doing that work, or if we are willing to move forward with manually invalidating backups 2. if there is a more simple/elegant way to do this that I'm not thinking of.

**Comment by taravancil** [comment]
I still need to implement this feature on VxScan, and I'm also waiting on some final design feedback. I'll leave it up to reviewers if they want to do another round of review now or wait until this is out of draft status.

**Comment by benadida** [comment]
> @benadida called out that manually invalidating the backup when we add/delete batches/sheets is brittle, because it requires us to remember to invalidate backups if/when we add new features. He made a good suggestion that we could compare export timestamps to the max of `batches.ended_at`, `sheets.created_at`, `sheets.deleted_at` to decide if a backup is up-to-date. I like this idea, but it still requires manually invalidating the export when we delete a batch.
> 
> We could get around this by setting `deleted_at` on batches instead of actually deleting batches (like we do for sheets), but before I move forward with that implementation, I wanted to pause and see if 1. we think it's worth doing that work, or if we are willing to move forward with manually invalidating backups 2. if there is a more simple/elegant way to do this that I'm not thinking of.

Unless we fully delete all ballot data, I think it makes sense to only mark things as deleted, rather than to actually delete them from the database. So I would suggest we do that, which ... does expand the scope here, but it is scope that is making our system better!

**Comment by eventualbuddha** [comment]
I agree that we should do soft deletes of batches. In fact, I'd suggest doing that as a precursor to this task that could be its own PR.

**Comment by taravancil** [comment]
> I agree that we should do soft deletes of batches. In fact, I'd suggest doing that as a precursor to this task that could be its own PR.

Good idea. I'll file a ticket.

File: frontends/bsd/src/screens/admin_actions_screen.tsx
Comment (beausmith) [reply frontends/bsd/src/screens/admin_actions_screen.tsx]: Oof. Perhaps place a `<br />` before the `<span>`? so that  the text is under the button and more tightly associated with the button? (sorry for the delay in response)

File: services/scan/src/store.ts
Comment (taravancil) [inline services/scan/src/store.ts:417]: I think this is the most important piece for reviewers to spend time on.

Specifically, if someone better at SQL is interested in helping me capture this heuristic in a single query, that would be awesome!

File: services/scan/src/store.ts
Comment (benadida) [inline services/scan/src/store.ts:446]: You could do this in two SQL statements, e.g.:

`select max(created_at) as maxCVrsCreatedAt, max(deleted_at) as maxCvrsDeletedAt from sheets`

File: services/scan/src/store.ts
Comment (benadida) [inline services/scan/src/store.ts:455]: possibly simplify by putting this logic in the `reduce`, e.g.:

`(max, curr) => ((max && max > curr) ?  max : curr),`

File: services/scan/src/store.ts
Comment (benadida) [reply services/scan/src/store.ts:417]: there's a way to do this in two simple SQL queries, I think, which, depending on your pov, might be ugly, or might be a nice encapsulation of the whole thing.

`select max(max(created_at, deleted_at)) <= ? from sheets`

`select max(max(created_at, deleted_at)) <= ? from batches`

If both are true, you're good to go.

**Review by benadida — APPROVED** [review]
This looks correct to me with one minor comment and one minor concern.

The minor comment: I think the code for `getCanUnconfigure` might be simplifiable a good bit, see my suggestion.

The minor concern: we're hammering sqlite on every `status` call which happens often, so doing a number of SQL queries every time gives me pause... it's not that much data so it's probably fine, but maybe worth a followup performance test ticket? (Also maybe worth an index on the timestamps in the `sheets` table.)

I'm good to move forward, though would suggest that followup performance check ticket, which is broader than just this ticket anyways since there is other code that, on `status`, is hammering sqlite.

File: services/scan/src/store.ts
Comment (taravancil) [reply services/scan/src/store.ts]: Good idea! Done.

File: services/scan/src/store.ts
Comment (taravancil) [reply services/scan/src/store.ts]: That doesn't handle if `curr` is `null`, so I'm going to keep using the `.filter` approach`. I think the intention that way is more clear, and since `n` is small I don't feel too bad about looping multiple times. We can and probably should revisit this in the followup perf ticket I'll file.

**Pull request merged**
